File: src/main/java/io/Cbuild/Cursor.java
package io.Cbuild;

import java.util.*;

public class Cursor {

    public static class CursorState {
        public int current;
        public int row;
        public int col;

        public CursorState() {}

        public CursorState(int current) {
            this.current = current;
        }

        public CursorState(int current, int row, int col) {
            this.current = current;
            this.row = row;
            this.col = col;
        }
    }

    public static class Floc {
        public int line;
        public Floc() { this.line = 0; }
        public Floc(int line) { this.line = line; }
    }



    public static final char END = '\0';


    public enum CharMask {
        End             (1L << 0),  // \0
        Blank           (1L << 1),  // space, tab
        Newline         (1L << 2),  // \n, \r
        Dollar          (1L << 3),  // $
        Semi            (1L << 4),  // ;
        Equals          (1L << 5),  // =
        Colon           (1L << 6),  // :
        Percent         (1L << 7),  // %
        Pipe            (1L << 8),  // |
        Dot             (1L << 9),  // .
        Comma           (1L << 10), // ,
        Escape          (1L << 11), // \
        Plus            (1L << 12), // +
        Minus           (1L << 13), // -
        Asterisk        (1L << 14), // *
        Slash           (1L << 15), // /
        DoubleQuote     (1L << 16), // "
        SingleQuote     (1L << 17), // '
        LeftBracket     (1L << 18), // [
        RightBracket    (1L << 19), // ]
        LeftParen       (1L << 20), // (
        RightParen      (1L << 21), // )
        LeftCurlyBrace  (1L << 22), // {
        RightCurlyBrace (1L << 23), // }
        UnderScore      (1L << 24), // _
        Hash            (1L << 25), // #
        Greater         (1L << 26), // >
        Less            (1L << 27), // <
        Bang            (1L << 28), // !
        Ampersand       (1L << 29), // &
        Caret           (1L << 30), // ^
        Question        (1L << 31), // ?
        Tilde           (1L << 32), // ~
        BackTick        (1L << 33), // `
        BackSlash       (1L << 34); // \

        public final long value;

        CharMask(long value) {
            this.value = value;
        }
    }


    private static final Map<Character, Long> STOPCHAR_MAP = new HashMap<>();

    static {
        STOPCHAR_MAP.put('\0', CharMask.End.value);

        STOPCHAR_MAP.put(' ',  CharMask.Blank.value);
        STOPCHAR_MAP.put('\t', CharMask.Blank.value);

        STOPCHAR_MAP.put('\n', CharMask.Newline.value);
        STOPCHAR_MAP.put('\r', CharMask.Newline.value);

        STOPCHAR_MAP.put('$',  CharMask.Escape.value | CharMask.Dollar.value);
        STOPCHAR_MAP.put(';',  CharMask.Semi.value);
        STOPCHAR_MAP.put('=',  CharMask.Equals.value);
        STOPCHAR_MAP.put(':',  CharMask.Colon.value);
        STOPCHAR_MAP.put('%',  CharMask.Percent.value);
        STOPCHAR_MAP.put('|',  CharMask.Pipe.value);
        STOPCHAR_MAP.put('.',  CharMask.Dot.value);
        STOPCHAR_MAP.put(',',  CharMask.Comma.value);
        STOPCHAR_MAP.put('\\', CharMask.Escape.value | CharMask.BackSlash.value);
        STOPCHAR_MAP.put('+',  CharMask.Plus.value);
        STOPCHAR_MAP.put('-',  CharMask.Minus.value);
        STOPCHAR_MAP.put('*',  CharMask.Asterisk.value);
        STOPCHAR_MAP.put('/',  CharMask.Slash.value);
        STOPCHAR_MAP.put('"',  CharMask.DoubleQuote.value);
        STOPCHAR_MAP.put('\'', CharMask.SingleQuote.value);

        STOPCHAR_MAP.put('[',  CharMask.LeftBracket.value);
        STOPCHAR_MAP.put(']',  CharMask.RightBracket.value);
        STOPCHAR_MAP.put('(',  CharMask.LeftParen.value);
        STOPCHAR_MAP.put(')',  CharMask.RightParen.value);
        STOPCHAR_MAP.put('{',  CharMask.LeftCurlyBrace.value);
        STOPCHAR_MAP.put('}',  CharMask.RightCurlyBrace.value);

        STOPCHAR_MAP.put('_',  CharMask.UnderScore.value);
        STOPCHAR_MAP.put('#',  CharMask.Hash.value);
        STOPCHAR_MAP.put('>',  CharMask.Greater.value);
        STOPCHAR_MAP.put('<',  CharMask.Less.value);
        STOPCHAR_MAP.put('!',  CharMask.Bang.value);
        STOPCHAR_MAP.put('&',  CharMask.Ampersand.value);
        STOPCHAR_MAP.put('^',  CharMask.Caret.value);
        STOPCHAR_MAP.put('?',  CharMask.Question.value);
        STOPCHAR_MAP.put('~',  CharMask.Tilde.value);
        STOPCHAR_MAP.put('`',  CharMask.BackTick.value);
    }

    public static final long EXPR_STOP =
            maskOf(
                    CharMask.End,
                    CharMask.Blank,
                    CharMask.Newline,
                    CharMask.Dollar,
                    CharMask.LeftParen,
                    CharMask.RightParen,
                    CharMask.LeftCurlyBrace,
                    CharMask.RightCurlyBrace,
                    CharMask.Colon,
                    CharMask.Comma,
                    CharMask.Equals,
                    CharMask.Hash,
                    CharMask.Pipe
            );

    public static final long RECIPE_STOP =
            maskOf(
                    CharMask.End,
                    CharMask.Newline,
                    CharMask.Dollar
            );

    public static final long ASSIGNMENT_LHS_STOP =
            maskOf(
                    CharMask.End,
                    CharMask.Blank,
                    CharMask.Newline,
                    CharMask.Equals,
                    CharMask.Colon,
                    CharMask.Plus,
                    CharMask.Question,
                    CharMask.Bang
            );

    public static final long COMMENT_START =
            CharMask.Hash.value;

    public static long getStopcharMask(char key) {
        return STOPCHAR_MAP.getOrDefault(key, (long)0);
    }


    public static long maskOr(CharMask a, CharMask b) {
        return a.value | b.value;
    }

    public static boolean anySet(long v, CharMask m) {
        return (v & m.value) != 0;
    }

    public static boolean anySet(long v, long mask) {
        return (v & mask) != 0;
    }

    public static boolean noneSet(long v, CharMask m) {
        return !anySet(v, m);
    }

    public static long maskOf(CharMask... masks) {
        long result = 0L;
        for (CharMask mask : masks) {
            result |= mask.value;
        }
        return result;
    }

    public static boolean noneSet(long v, long mask) {
        return !anySet(v, mask);
    }

    public static boolean stopSet(char c, CharMask m) {
        return anySet(getStopcharMask(c), m);
    }

    public static boolean stopSet(char c, long mask) {
        return anySet(getStopcharMask(c), mask);
    }

    public static boolean isBlank(char c) {
        return stopSet(c, CharMask.Blank);
    }

    public static boolean isSpace(char c) {
        return stopSet(c, maskOr(CharMask.Newline, CharMask.Blank));
    }


    public static char peek(String text, CursorState cursor) {
        if (cursor.current >= text.length()) return END;
        return text.charAt(cursor.current);
    }

    public static char advance(String text, CursorState cursor) {
        if (cursor.current >= text.length()) return END;
        return text.charAt(cursor.current++);
    }

    public static char peekNext(String text, CursorState cursor) {
        if (cursor.current + 1 >= text.length()) return END;
        return text.charAt(cursor.current + 1);
    }

    public static boolean match(String text, CursorState cursor, char expected) {
        if (cursor.current >= text.length()) return false;
        if (text.charAt(cursor.current) == expected) {
            cursor.current++;
            return true;
        }
        return false;
    }

    public static void consumeBlank(String text, CursorState cursor) {
        while (isBlank(peek(text, cursor))) cursor.current++;
    }

    public static void consumeSpace(String text, CursorState cursor) {
        while (isSpace(peek(text, cursor))) cursor.current++;
    }

    public static class Pchar {
        public enum EscapeType {
            BackSlash,
            Dollar
        }

        public int row;
        public int col;
        public char c;
        public boolean isEscaped;
        public EscapeType escapeType;

        public Pchar(char c) {
            this(0, 0, c, false);
        }

        public Pchar(char c, boolean isEscaped) {
            this(0, 0, c, isEscaped);
        }

        public Pchar(int row, int col, char c) {
            this(row, col, c, false);
        }

        public Pchar(int row, int col, char c, boolean isEscaped) {
            this.row = row;
            this.col = col;
            this.c = c;
            this.isEscaped = isEscaped;
        }

        public Pchar(int row,
                     int col,
                     char c,
                     boolean isEscaped,
                     EscapeType escapeType) {
            this.row = row;
            this.col = col;
            this.c = c;
            this.isEscaped = isEscaped;
            this.escapeType = escapeType;
        }
    }

    private static final Pchar EOF_PCHAR = new Pchar(END);

    public static boolean stopSet(Pchar p, CharMask m) {
        return stopSet(p.c, m);
    }

    public static boolean isBlank(Pchar p) {
        return isBlank(p.c);
    }

    public static boolean isSpace(Pchar p) {
        return isSpace(p.c);
    }

    public static Pchar peek(List<Pchar> buf, int current) {
        if (current >= buf.size()) return EOF_PCHAR;
        return buf.get(current);
    }

    public static Pchar peekNext(List<Pchar> buf, int current) {
        if (current + 1 >= buf.size()) return EOF_PCHAR;
        return buf.get(current + 1);
    }

    public static Pchar advance(List<Pchar> buf, CursorState cursor) {
        Pchar p = peek(buf, cursor.current);
        if (cursor.current < buf.size()) cursor.current++;
        return p;
    }

    public static Pchar prev(List<Pchar> buf, int current) {
        if (current - 1 < 0) return EOF_PCHAR;
        return buf.get(current - 1);
    }

    public static Pchar prev(List<Pchar> buf, CursorState cursorState) {
        return prev(buf, cursorState.current);
    }

    public static char prevChar(List<Pchar> buf, int current) {
        return prev(buf, current).c;
    }

    public static char prev(String text, CursorState cursor) {
        if (cursor.current - 1 < 0) return END;
        return text.charAt(cursor.current - 1);
    }

    public static char peekChar(List<Pchar> buf, int current) {
        return peek(buf, current).c;
    }

    public static char peekNextChar(List<Pchar> buf, int current) {
        return peekNext(buf, current).c;
    }

    public static char advanceChar(List<Pchar> buf, CursorState cursor) {
        return advance(buf, cursor).c;
    }

    public static boolean match(List<Pchar> buf, CursorState cursor, char expected) {
        if (peek(buf, cursor.current).c == expected) {
            cursor.current++;
            return true;
        }
        return false;
    }

    public static void consumeBlank(List<Pchar> buf, CursorState cursor) {
        while (isBlank(peek(buf, cursor.current))) cursor.current++;
    }

    public static void consumeSpace(List<Pchar> buf, CursorState cursor) {
        while (isSpace(peek(buf, cursor.current))) cursor.current++;
    }


    public static List<Pchar> mergePBuffer(List<Pchar> buf1, List<Pchar> buf2) {
        if (buf1.isEmpty()) return new ArrayList<>(buf2);
        if (buf2.isEmpty()) return new ArrayList<>(buf1);
        List<Pchar> result = new ArrayList<>(buf1.size() + buf2.size());
        result.addAll(buf1);
        result.addAll(buf2);
        return result;
    }

    public static List<Pchar> toPBuffer(String str, int line) {
        List<Pchar> buf = new ArrayList<>(str.length());
        for (int i = 0; i < str.length(); i++) {
            buf.add(new Pchar(line, i + 1, str.charAt(i)));
        }
        return buf;
    }

    public static String pBufferToString(List<Pchar> buf) {
        StringBuilder sb = new StringBuilder(buf.size());
        for (Pchar p : buf) sb.append(p.c);
        return sb.toString();
    }

    public static void printPBuffer(List<Pchar> buf) {
        for (Pchar p : buf) System.out.print(p.c);
    }

    public static boolean isEscaped(String text, int charIndex, char escapeCharacter) {
        int i = charIndex - 1;
        int counter = 0;
        while (i >= 0 && text.charAt(i) == escapeCharacter) {
            i--;
            counter++;
        }
        return counter % 2 == 1;
    }

    public static boolean isEscapedBackslash(String text, int charIndex) {
        return isEscaped(text, charIndex, '\\');
    }

    public static boolean isEscaped(List<Cursor.Pchar> buf, int charIndex, char escapeCharacter) {
        int i = charIndex - 1;
        int counter = 0;
        while (i >= 0 && buf.get(i).c == escapeCharacter) {
            i--;
            counter++;
        }
        return counter % 2 == 1;
    }

    public static boolean isEscapedBackslash(List<Cursor.Pchar> buf, int charIndex) {
        return isEscaped(buf, charIndex, '\\');
    }

}

File: src/Cursor.ts
declare const process: { stdout: { write(value: string): void } };

export class Cursor {
  public static readonly END: string = "\0";

  private static get STOPCHAR_MAP(): Map<string, bigint> {
    return new Map<string, bigint>([
      ["\0", Cursor.CharMask.End],

      [" ", Cursor.CharMask.Blank],
      ["\t", Cursor.CharMask.Blank],

      ["\n", Cursor.CharMask.Newline],
      ["\r", Cursor.CharMask.Newline],

      ["$", Cursor.CharMask.Escape | Cursor.CharMask.Dollar],
      [";", Cursor.CharMask.Semi],
      ["=", Cursor.CharMask.Equals],
      [":", Cursor.CharMask.Colon],
      ["%", Cursor.CharMask.Percent],
      ["|", Cursor.CharMask.Pipe],
      [".", Cursor.CharMask.Dot],
      [",", Cursor.CharMask.Comma],
      ["\\", Cursor.CharMask.Escape | Cursor.CharMask.BackSlash],
      ["+", Cursor.CharMask.Plus],
      ["-", Cursor.CharMask.Minus],
      ["*", Cursor.CharMask.Asterisk],
      ["/", Cursor.CharMask.Slash],
      ['"', Cursor.CharMask.DoubleQuote],
      ["'", Cursor.CharMask.SingleQuote],

      ["[", Cursor.CharMask.LeftBracket],
      ["]", Cursor.CharMask.RightBracket],
      ["(", Cursor.CharMask.LeftParen],
      [")", Cursor.CharMask.RightParen],
      ["{", Cursor.CharMask.LeftCurlyBrace],
      ["}", Cursor.CharMask.RightCurlyBrace],

      ["_", Cursor.CharMask.UnderScore],
      ["#", Cursor.CharMask.Hash],
      [">", Cursor.CharMask.Greater],
      ["<", Cursor.CharMask.Less],
      ["!", Cursor.CharMask.Bang],
      ["&", Cursor.CharMask.Ampersand],
      ["^", Cursor.CharMask.Caret],
      ["?", Cursor.CharMask.Question],
      ["~", Cursor.CharMask.Tilde],
      ["`", Cursor.CharMask.BackTick],
    ]);
  }

  public static get EXPR_STOP(): bigint {
    return Cursor.maskOf(
      Cursor.CharMask.End,
      Cursor.CharMask.Blank,
      Cursor.CharMask.Newline,
      Cursor.CharMask.Dollar,
      Cursor.CharMask.LeftParen,
      Cursor.CharMask.RightParen,
      Cursor.CharMask.LeftCurlyBrace,
      Cursor.CharMask.RightCurlyBrace,
      Cursor.CharMask.Colon,
      Cursor.CharMask.Comma,
      Cursor.CharMask.Equals,
      Cursor.CharMask.Hash,
      Cursor.CharMask.Pipe,
    );
  }

  public static get RECIPE_STOP(): bigint {
    return Cursor.maskOf(
      Cursor.CharMask.End,
      Cursor.CharMask.Newline,
      Cursor.CharMask.Dollar,
    );
  }

  public static get ASSIGNMENT_LHS_STOP(): bigint {
    return Cursor.maskOf(
      Cursor.CharMask.End,
      Cursor.CharMask.Blank,
      Cursor.CharMask.Newline,
      Cursor.CharMask.Equals,
      Cursor.CharMask.Colon,
      Cursor.CharMask.Plus,
      Cursor.CharMask.Question,
      Cursor.CharMask.Bang,
    );
  }

  public static get COMMENT_START(): bigint {
    return Cursor.CharMask.Hash;
  }

  private static get EOF_PCHAR(): Cursor.Pchar {
    return EOF_PCHAR_VALUE;
  }

  public static getStopcharMask(key: string): bigint {
    return Cursor.STOPCHAR_MAP.get(key) ?? 0n;
  }

  public static maskOr(a: Cursor.CharMask, b: Cursor.CharMask): bigint {
    return a | b;
  }

  public static anySet(v: bigint, m: Cursor.CharMask): boolean;
  public static anySet(v: bigint, mask: bigint): boolean;
  public static anySet(v: bigint, mask: bigint): boolean {
    return (v & mask) !== 0n;
  }

  public static noneSet(v: bigint, m: Cursor.CharMask): boolean;
  public static noneSet(v: bigint, mask: bigint): boolean;
  public static noneSet(v: bigint, mask: bigint): boolean {
    return !Cursor.anySet(v, mask);
  }

  public static maskOf(...masks: Cursor.CharMask[]): bigint {
    let result = 0n;
    for (const mask of masks) {
      result |= mask;
    }
    return result;
  }

  public static stopSet(c: string, m: Cursor.CharMask): boolean;
  public static stopSet(c: string, mask: bigint): boolean;
  public static stopSet(p: Cursor.Pchar, m: Cursor.CharMask): boolean;
  public static stopSet(cOrP: string | Cursor.Pchar, mask: bigint): boolean {
    const c = typeof cOrP === "string" ? cOrP : cOrP.c;
    return Cursor.anySet(Cursor.getStopcharMask(c), mask);
  }

  public static isBlank(c: string): boolean;
  public static isBlank(p: Cursor.Pchar): boolean;
  public static isBlank(cOrP: string | Cursor.Pchar): boolean {
    const c = typeof cOrP === "string" ? cOrP : cOrP.c;
    return Cursor.stopSet(c, Cursor.CharMask.Blank);
  }

  public static isSpace(c: string): boolean;
  public static isSpace(p: Cursor.Pchar): boolean;
  public static isSpace(cOrP: string | Cursor.Pchar): boolean {
    const c = typeof cOrP === "string" ? cOrP : cOrP.c;
    return Cursor.stopSet(
      c,
      Cursor.maskOr(Cursor.CharMask.Newline, Cursor.CharMask.Blank),
    );
  }

  public static peek(text: string, cursor: Cursor.CursorState): string;
  public static peek(buf: Cursor.Pchar[], current: number): Cursor.Pchar;
  public static peek(
    textOrBuf: string | Cursor.Pchar[],
    cursorOrCurrent: Cursor.CursorState | number,
  ): string | Cursor.Pchar {
    if (typeof textOrBuf === "string") {
      const cursor = cursorOrCurrent as Cursor.CursorState;
      if (cursor.current >= textOrBuf.length) return Cursor.END;
      return textOrBuf.charAt(cursor.current);
    }
    const current = cursorOrCurrent as number;
    if (current >= textOrBuf.length) return Cursor.EOF_PCHAR;
    return textOrBuf[current];
  }

  public static advance(text: string, cursor: Cursor.CursorState): string;
  public static advance(
    buf: Cursor.Pchar[],
    cursor: Cursor.CursorState,
  ): Cursor.Pchar;
  public static advance(
    textOrBuf: string | Cursor.Pchar[],
    cursor: Cursor.CursorState,
  ): string | Cursor.Pchar {
    if (typeof textOrBuf === "string") {
      if (cursor.current >= textOrBuf.length) return Cursor.END;
      return textOrBuf.charAt(cursor.current++);
    }
    const p = Cursor.peek(textOrBuf, cursor.current);
    if (cursor.current < textOrBuf.length) cursor.current++;
    return p;
  }

  public static peekNext(text: string, cursor: Cursor.CursorState): string;
  public static peekNext(buf: Cursor.Pchar[], current: number): Cursor.Pchar;
  public static peekNext(
    textOrBuf: string | Cursor.Pchar[],
    cursorOrCurrent: Cursor.CursorState | number,
  ): string | Cursor.Pchar {
    if (typeof textOrBuf === "string") {
      const cursor = cursorOrCurrent as Cursor.CursorState;
      if (cursor.current + 1 >= textOrBuf.length) return Cursor.END;
      return textOrBuf.charAt(cursor.current + 1);
    }
    const current = cursorOrCurrent as number;
    if (current + 1 >= textOrBuf.length) return Cursor.EOF_PCHAR;
    return textOrBuf[current + 1];
  }

  public static match(
    text: string,
    cursor: Cursor.CursorState,
    expected: string,
  ): boolean;
  public static match(
    buf: Cursor.Pchar[],
    cursor: Cursor.CursorState,
    expected: string,
  ): boolean;
  public static match(
    textOrBuf: string | Cursor.Pchar[],
    cursor: Cursor.CursorState,
    expected: string,
  ): boolean {
    if (typeof textOrBuf === "string") {
      if (cursor.current >= textOrBuf.length) return false;
      if (textOrBuf.charAt(cursor.current) === expected) {
        cursor.current++;
        return true;
      }
      return false;
    }
    if (Cursor.peek(textOrBuf, cursor.current).c === expected) {
      cursor.current++;
      return true;
    }
    return false;
  }

  public static consumeBlank(text: string, cursor: Cursor.CursorState): void;
  public static consumeBlank(
    buf: Cursor.Pchar[],
    cursor: Cursor.CursorState,
  ): void;
  public static consumeBlank(
    textOrBuf: string | Cursor.Pchar[],
    cursor: Cursor.CursorState,
  ): void {
    if (typeof textOrBuf === "string") {
      while (Cursor.isBlank(Cursor.peek(textOrBuf, cursor))) cursor.current++;
    } else {
      while (Cursor.isBlank(Cursor.peek(textOrBuf, cursor.current)))
        cursor.current++;
    }
  }

  public static consumeSpace(text: string, cursor: Cursor.CursorState): void;
  public static consumeSpace(
    buf: Cursor.Pchar[],
    cursor: Cursor.CursorState,
  ): void;
  public static consumeSpace(
    textOrBuf: string | Cursor.Pchar[],
    cursor: Cursor.CursorState,
  ): void {
    if (typeof textOrBuf === "string") {
      while (Cursor.isSpace(Cursor.peek(textOrBuf, cursor))) cursor.current++;
    } else {
      while (Cursor.isSpace(Cursor.peek(textOrBuf, cursor.current)))
        cursor.current++;
    }
  }

  public static prev(buf: Cursor.Pchar[], current: number): Cursor.Pchar;
  public static prev(
    buf: Cursor.Pchar[],
    cursorState: Cursor.CursorState,
  ): Cursor.Pchar;
  public static prev(text: string, cursor: Cursor.CursorState): string;
  public static prev(
    bufOrText: Cursor.Pchar[] | string,
    currentOrCursor: number | Cursor.CursorState,
  ): Cursor.Pchar | string {
    if (typeof bufOrText === "string") {
      const cursor = currentOrCursor as Cursor.CursorState;
      if (cursor.current - 1 < 0) return Cursor.END;
      return bufOrText.charAt(cursor.current - 1);
    }
    const current =
      typeof currentOrCursor === "number"
        ? currentOrCursor
        : currentOrCursor.current;
    if (current - 1 < 0) return Cursor.EOF_PCHAR;
    return bufOrText[current - 1];
  }

  public static prevChar(buf: Cursor.Pchar[], current: number): string {
    return Cursor.prev(buf, current).c;
  }

  public static peekChar(buf: Cursor.Pchar[], current: number): string {
    return Cursor.peek(buf, current).c;
  }

  public static peekNextChar(buf: Cursor.Pchar[], current: number): string {
    return Cursor.peekNext(buf, current).c;
  }

  public static advanceChar(
    buf: Cursor.Pchar[],
    cursor: Cursor.CursorState,
  ): string {
    return Cursor.advance(buf, cursor).c;
  }

  public static mergePBuffer(
    buf1: Cursor.Pchar[],
    buf2: Cursor.Pchar[],
  ): Cursor.Pchar[] {
    if (buf1.length === 0) return [...buf2];
    if (buf2.length === 0) return [...buf1];
    return [...buf1, ...buf2];
  }

  public static toPBuffer(str: string, line: number): Cursor.Pchar[] {
    const buf: Cursor.Pchar[] = [];
    for (let i = 0; i < str.length; i++) {
      buf.push(new Cursor.Pchar(line, i + 1, str.charAt(i)));
    }
    return buf;
  }

  public static pBufferToString(buf: Cursor.Pchar[]): string {
    let result = "";
    for (const p of buf) result += p.c;
    return result;
  }

  public static printPBuffer(buf: Cursor.Pchar[]): void {
    for (const p of buf) process.stdout.write(p.c);
  }

  public static isEscaped(
    text: string,
    charIndex: number,
    escapeCharacter: string,
  ): boolean;
  public static isEscaped(
    buf: Cursor.Pchar[],
    charIndex: number,
    escapeCharacter: string,
  ): boolean;
  public static isEscaped(
    textOrBuf: string | Cursor.Pchar[],
    charIndex: number,
    escapeCharacter: string,
  ): boolean {
    let i = charIndex - 1;
    let counter = 0;
    while (
      i >= 0 &&
      (typeof textOrBuf === "string" ? textOrBuf.charAt(i) : textOrBuf[i].c) ===
        escapeCharacter
    ) {
      i--;
      counter++;
    }
    return counter % 2 === 1;
  }

  public static isEscapedBackslash(text: string, charIndex: number): boolean;
  public static isEscapedBackslash(
    buf: Cursor.Pchar[],
    charIndex: number,
  ): boolean;
  public static isEscapedBackslash(
    textOrBuf: string | Cursor.Pchar[],
    charIndex: number,
  ): boolean {
    if (typeof textOrBuf === "string")
      return Cursor.isEscaped(textOrBuf, charIndex, "\\");
    return Cursor.isEscaped(textOrBuf, charIndex, "\\");
  }
}

export namespace Cursor {
  export class CursorState {
    public current: number;
    public row: number;
    public col: number;
    public processedRow: number;
    public processedCol: number;

    public constructor();
    public constructor(current: number);
    public constructor(current: number, row: number, col: number);
    public constructor(current = 0, row = 0, col = 0) {
      this.current = current;
      this.row = row;
      this.col = col;
      this.processedRow = 0;
      this.processedCol = 0;
    }
  }

  export class Floc {
    public line: number;

    public constructor();
    public constructor(line: number);
    public constructor(line = 0) {
      this.line = line;
    }
  }

  export const CharMask = {
    End: 1n << 0n,
    Blank: 1n << 1n,
    Newline: 1n << 2n,
    Dollar: 1n << 3n,
    Semi: 1n << 4n,
    Equals: 1n << 5n,
    Colon: 1n << 6n,
    Percent: 1n << 7n,
    Pipe: 1n << 8n,
    Dot: 1n << 9n,
    Comma: 1n << 10n,
    Escape: 1n << 11n,
    Plus: 1n << 12n,
    Minus: 1n << 13n,
    Asterisk: 1n << 14n,
    Slash: 1n << 15n,
    DoubleQuote: 1n << 16n,
    SingleQuote: 1n << 17n,
    LeftBracket: 1n << 18n,
    RightBracket: 1n << 19n,
    LeftParen: 1n << 20n,
    RightParen: 1n << 21n,
    LeftCurlyBrace: 1n << 22n,
    RightCurlyBrace: 1n << 23n,
    UnderScore: 1n << 24n,
    Hash: 1n << 25n,
    Greater: 1n << 26n,
    Less: 1n << 27n,
    Bang: 1n << 28n,
    Ampersand: 1n << 29n,
    Caret: 1n << 30n,
    Question: 1n << 31n,
    Tilde: 1n << 32n,
    BackTick: 1n << 33n,
    BackSlash: 1n << 34n,
  } as const;

  export type CharMask = (typeof CharMask)[keyof typeof CharMask];

  export class Pchar {
    public row: number;
    public col: number;
    public processedRow: number;
    public processedCol: number;
    public c: string;
    public isEscaped: boolean;
    public escapeType: Pchar.EscapeType | undefined;
    public context: Pchar.Context | undefined;

    public constructor(c: string);
    public constructor(c: string, isEscaped: boolean);
    public constructor(row: number, col: number, c: string);
    public constructor(row: number, col: number, c: string, isEscaped: boolean);
    public constructor(
      row: number,
      col: number,
      c: string,
      isEscaped: boolean,
      escapeType: Pchar.EscapeType,
      context: Pchar.Context,
    );
    public constructor(
      rowOrC: number | string,
      colOrEscaped: number | boolean = 0,
      c = Cursor.END,
      isEscaped = false,
      escapeType?: Pchar.EscapeType,
      context?: Pchar.Context,
    ) {
      if (typeof rowOrC === "string") {
        this.row = 0;
        this.col = 0;
        this.c = rowOrC;
        this.isEscaped =
          typeof colOrEscaped === "boolean" ? colOrEscaped : false;
      } else {
        this.row = rowOrC;
        this.col = colOrEscaped as number;
        this.c = c;
        this.isEscaped = isEscaped;
      }
      this.processedRow = this.row;
      this.processedCol = this.col;
      this.escapeType = escapeType;
      this.context = context;
    }

    public toString(): string {
      return `row=${this.row}, col=${this.col}, processedRow=${this.processedRow}, processedCol=${this.processedCol}, c='${this.c}', isEscaped=${this.isEscaped}, escapeType=${Pchar.EscapeType[this.escapeType as Pchar.EscapeType]}, context=${Pchar.Context[this.context as Pchar.Context]}`;
    }
  }

  export namespace Pchar {
    export enum Context {
      YSHARP_HOOK,
      LUA_HOOK,
      CBUILD_HOOK,
      NORMAL,
    }

    export enum EscapeType {
      BackSlash,
      Dollar,
    }
  }
}

const EOF_PCHAR_VALUE = new Cursor.Pchar(Cursor.END);
